### ISSUE TITLE
Getting token from Cloud Shell intermittently fails with 400 Client Error: Bad Request

### ISSUE THREAD

**Issue by Kalyan-Alamuru** [Azure/azure-cli#11749]
I'm getting following when I'm running following command : 

ARM_CLIENT_SECRET=$(az ad sp create-for-rbac \
  --name http://tf-sp-$UNIQUE_ID \
  --role Contributor \
  --scopes "/subscriptions/$ARM_SUBSCRIPTION_ID" \
  --query password \
  --output tsv)

Please note that I've stored ARM Subscription ID successfully and ran above command as part of creating Service Principal.

## Describe the bug

**Command Name**
`az ad sp create-for-rbac`

**Errors:**
```
400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token
Traceback (most recent call last):
python3.6/site-packages/knack/cli.py, ln 206, in invoke
    cmd_result = self.invocation.execute(args)
cli/core/commands/__init__.py, ln 608, in execute
    raise ex
cli/core/commands/__init__.py, ln 666, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
...
python3.6/site-packages/msrestazure/azure_active_directory.py, ln 486, in get_msi_token
    result.raise_for_status()
python3.6/site-packages/requests/models.py, ln 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token
```

## To Reproduce:
Steps to reproduce the behavior. Note that argument values have been redacted, as they may contain sensitive information.

- _Put any pre-requisite steps here..._
- `az ad sp create-for-rbac --name {} --role {} --scopes {} --query {} --output {}`

## Environment Summary
```
Linux-4.15.0-1064-azure-x86_64-with-debian-stretch-sid
Python 3.6.5
Shell: bash

azure-cli 2.0.78

```

**Comment by bim-msft** [comment]
@jiasli Please take a look.

**Comment by jiasli** [comment]
This is Cloud Shell issue. Could you run with `--debug` and share the output? 

To get unblocked, please run `az login` and retry the command.

**Comment by Kalyan-Alamuru** [comment]
I reran the module again and it worked fine, Only difference this time is I ran Terraform Destroy command to delete the plan and then ran az ad sp create command and it worked fine, It doesn't really explain what happened !!!

**Comment by jiasli** [comment]
@Kalyan-Alamuru, we do see random 400 error from Cloud Shell. We are working with Cloud Shell team on this issue. 

If the issue happens again, please provide `--debug` output in this thread. Thanks.

**Comment by hubert-associates** [comment]
> This is Cloud Shell issue. Could you run with `--debug` and share the output?
> 
> To get unblocked, please run `az login` and retry the command.

This worked for me. Thanks.

**Comment by maertendMSFT** [comment]
This should be fixed now.  I cannot repro, please close.

**Comment by jiasli** [comment]
> This should be fixed now. I cannot repro, please close.

Hi @maertendMSFT, this issue doesn't happen **consistently**, but **intermittently**. If there are any changes on the Cloud Shell side, could you put some more details?

**Comment by maertendMSFT** [comment]
@jiasli , this appears to fundamentally be the same issue as a few of the other issues that I requested to close, all of which are solved with a `az login`.  We have put out a fix and that should have been included in the latest release (20200702) and be available for all customers now.

**Comment by yonzhan** [comment]
@maertendMSFT  good news, please let us know when the latest change rolls out.

**Comment by maertendMSFT** [comment]
The updated image should be available everywhere already :)

**Comment by digimaun** [comment]
I wouldn't close this issue until there is more confirmation. The issue still happens in cloudshell (today).

**Comment by yonzhan** [comment]
@maertendMSFT  any update for this issue ? There are a bunch of customers are impacted by this.

**Comment by AgSync-Aaron** [comment]
This still appears to be an issue. Well it's an issue for me anyway. I hope these logs can help.
```
aaron@Azure:~$ az keyvault secret list-versions --vault-name $vault_name --name $vault_cert_name
Error occurred in request., HTTPError: 400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token
```
Here is the tail end of the output with the `--debug` flag
```
urllib3.connectionpool : Starting new HTTP connection (1): localhost:50342
urllib3.connectionpool : http://localhost:50342 "POST /oauth2/token HTTP/1.1" 400 121
msrestazure.azure_active_directory : MSI: Retrieving a token from http://localhost:50342/oauth2/token, with payload {'resource': 'https://vault.azure.net'}
msrestazure.azure_active_directory : MSI: Failed to retrieve a token from 'http://localhost:50342/oauth2/token' with an error of '400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token'. This could be caused by the MSI extension not yet fully provisioned.
msrest.exceptions : Error occurred in request., HTTPError: 400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token
cli.azure.cli.core.util : Error occurred in request., HTTPError: 400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token
Error occurred in request., HTTPError: 400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token
Event: Cli.PostExecute [<function AzCliLogging.deinit_cmd_metadata_logging at 0x7f8bf399ad90>]
az_command_data_logger : exit code: 1
telemetry.save : Save telemetry record of length 5447 in cache
telemetry.check : Negative: The /home/aaron/.azure/telemetry.txt was modified at 2020-08-11 21:39:07.906455, which in less than 600.000000 s
Command ran in 40.650 seconds (init: 0.041, invoke: 40.610)
```

**Comment by haitch** [comment]
@jiasli  this is still happening and hurt AKS user experience, please prioritize and fix this issue.

**Comment by giggio** [comment]
I'm also getting this error. For me, it was when I ran `az storage blob generate-sas`.
Running `az login` fixed it.
I'm on Windows Terminal connecting directly to the cloud shell.

**Comment by oguzhanf** [comment]
I get the same error using Windows 11 - Windows Terminal, click on the drop-down to get to an Azure Cli instance. Following the device login page I'm able to work other commands but not the below: 

```
az ad user create --display-name "..." --password "..." --user-principal-name "..."

Failed to connect to MSI. Please make sure MSI is configured correctly.
Get Token request returned: <Response [400]>
```

**Comment by jiasli** [comment]
@sherdana, your message is corrupted. Also, you are not on Cloud Shell, but Windows machine. Please create a new issue with detailed information and error message.

**Comment by sabbour** [comment]
Running `az login` isn't an acceptable workaround. MSI login allows for elevated commands like "az ad app" which will be blocked otherwise.

**Comment by xfz11** [comment]
This issue doesn't happen on Cloud Shell created by Azure Portal, but still exists on Cloud shell created by Doc site. Is there anyone can connect Cloud Shell team to investigate the issue?

**Comment by luiznazari** [comment]
A similar error is occurring in the Microsoft Training lab [Exercise - Configure a system-assigned managed identity for an Azure VM](https://learn.microsoft.com/en-us/training/modules/authenticate-apps-with-managed-identities/5-exercise-configure-managed-identity-for-vm), first command. 
![image](https://user-images.githubusercontent.com/7645409/235321068-48b97818-8953-4cc5-bb3e-7b7a0978b6a5.png)

**Comment by MikeKemmerer** [comment]
I, too, am still having this issue. The workaround of using `az login` does not work in my tenant.

**Comment by CornerstoneII** [comment]
`Invoking `az ad app permission grant --id 54e27600-df4f-4e97-96b1-xxxxxxxxxxxx --api 00000003-0000-0000-c000-000000000000` is needed to make the change effective
Failed to connect to MSI. Please make sure MSI is configured correctly.
Get Token request returned: <Response [400]>`

**Comment by CornerstoneII** [comment]
> `400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token` is a known issue of Cloud Shell that it intermittently fails with this error.
> 
> ## Workarounds
> There are 2 workarounds:
> 
> 1. Use Azure CLI on a local machine
> 2. In Cloud Shell, run `az login` and retry the command

This workaround no longer works!!!

> `400 Client Error: Bad Request for url: http://localhost:50342/oauth2/token` is a known issue of Cloud Shell that it intermittently fails with this error.
> 
> ## Workarounds
> There are 2 workarounds:
> 
> 1. Use Azure CLI on a local machine
> 2. In Cloud Shell, run `az login` and retry the command

The workaround for cloudshell does not work!!!

`
Invoking `az ad app permission grant --id 54e27600-df4f-4e97-96b1-6aab7c1e0189 --api 00000003-0000-0000-c000-000000000000` is needed to make the change effective
Failed to connect to MSI. Please make sure MSI is configured correctly.
Get Token request returned: <Response [400]>
`

**Comment by MantisTree** [comment]
This is still an issue that crashed my deployment.  running the following command without `az login`:

`az aks create --resource-group ClusterRG01 --name AKSCluster02 --node-count 2 --enable-addons monitoring --generate-ssh-keys`

resulted in error:

`failed to connect to MSI. Please make sure MSI is configured correctly.`

but after running `az login` in cloud shell and retrying the same command (with cluster number increment) succeeded without any error
